opt/scopeserver/dataserver/utils/search.py

```diff
@@ -0,0 +1,323 @@
+from scopeserver.dataserver.utils.search_space import SearchSpace
+from scopeserver.dataserver.utils import data_file_handler as dfh
+
+logger = logging.getLogger(__name__)
+
+# casefold feature, original feature, feature type
+SearchMatch = NamedTuple("SearchResult", [("original_feature", str), ("resolved_feature", str), ("feature_type", str)])
+
+
+DEFINED_SEARCH_TYPES = {
+    "region_gene_link": {
+        "final_category": "gene",
+        "join_text_single": "is linked to",
+        "join_text_multiple": "are linked to",
+    },
+    "regulon_target": {
+        "final_category": "regulon",
+        "join_text_single": "is targeted by",
+        "join_text_multiple": "are targeted by",
+    },
+    "annotation_category": {"final_category": "annotation", "join_text_single": "is a", "join_text_multiple": "are"},
+    "marker_gene": {
+        "final_category": "cluster_category",
+        "join_text_single": "is a marker of",
+        "join_text_multiple": "are markers of",
+    },
+    "cluster_annotation": {
+        "final_category": "cluster_category",
+        "join_text_single": "is a suggested annotation of",
+        "join_text_multiple": "are suggested annotations of",
+    },
+}
+
+
+def determine_species_parameters(
+    loom: Loom, search_term: str
+) -> Tuple[SearchSpace, str, Union[constant.Species, None]]:
+    """Identify the species in the case of a cross species search
+
+    Args:
+        loom (Loom): Loom object
+        search_term (str): Term the user has searched for
+
+    Returns:
+        Tuple[SearchSpace, str, Union[constant.Species, None]]: 
+        Return the correct (cross-species or not) search space, 
+        the trimmed search term and the species being searched for.
+    """
+    if search_term.startswith("hsap\\"):
+        search_space = loom.hsap_ss
+        cross_species = constant.Species.HSAP
+        search_term = search_term[5:]
+    elif search_term.startswith("mmus\\"):
+        search_space = loom.mmus_ss
+        cross_species = constant.Species.MMUS
+        search_term = search_term[5:]
+    else:
+        search_space = loom.ss
+        cross_species = None
+    return search_space, search_term, cross_species
+
+
+def find_matches(search_term: str, casefold_term: str, search_space: SearchSpace) -> List[SearchMatch]:
+    """
+    Search for matches in the search space.
+
+    Find keys in the search space where the search term matches (casefolded)
+    the searchable element. Return the match, the matches category and relationship.
+
+    Args:
+        search_term (str): Term the user typed
+        casefold_term (str): Casefolded term
+        search_space (SearchSpace): Search space from loom object
+
+    Returns:
+        List[SearchMatch]: A sorted list of the matches to the users search term
+    """
+
+    if len(casefold_term) == 1:
+        matches = [x for x in search_space if casefold_term == x[0]]
+    else:
+        matches = [x for x in search_space if casefold_term in x[0]]
```

Comment:
My code checks to see if a single letter was searched for, if so, we limit the search to only provide exact matches, otherwise searches take forever and return thousands of values

Reply:
Then the results need to be ranked (a la the `match_result_cost` function I defined in a later suggestion) and the top N "best" matches returned.

Reply:
So this line is actually the most expensive operation and it increases with the number of results. For example, in one file, searching for `c` with just `in` (which matches 2898830 entries) takes ~4.5 seconds to find them and sort them, but if I use `==` for length 1 terms, this matches 2 results and takes ~6ms

Reply:
Not sure how helpful it would be for SCope, but pandas contains str functions that works on numpy arrays directly, which might be useful to find matches faster: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.contains.html

Reply:
OK that's very interesting. `2_898_830` isn't a particularly big number. I'm curious if making `matches` a generator would change the performance at all:
```python3
matches = (x for x in search_space if casefold_term in x[0])
```

@ghuls idea to attack it with numpy is good in principle. A naive approach would be to transform the search space into an `ndarray` then filter that:
```python3
ss = np.array(search_space)
matches = ss[np.char.find(ss, casefold_term) != -1]
```

I measured it like so:
```python3
search_space = np.array([str(x) for x in range(3_000_000)])
timeit.timeit("search_space[np.char.find(search_space, '2') != -1]", number=100, globals={'search_space': search_space, 'np': np})
```
Each search takes ~1.48 seconds without sorting so the naive approach is no good either.

Alternatively using a `set()` might improve things:
```python3
search_space = set((str(x) for x in range(3_000_000)))
timeit.timeit("[x for x in search_space if '2' in x]", number=100, globals={'search_space': search_space})
```
Down to ~0.58 seconds without sorting. Not bad. Not 6ms.

Reply:
For now, I'm keeping the exception for 1 character searches, but switching to a generator did speed that step up also